### PR TITLE
fix: confirmation when transaction (STX) is pending cross-network

### LINKED ISSUE #25579: [Bug]: When a user has a Smart transaction in a Pending state and initiates a transaction through a dApp on a different network, the dApp transaction is generated with an empty nonce. Pressing the 'Confirm' button then leads to an endless loading spinner.
### Describe the bug

 v12.0.0 build [https://github.com/MetaMask/metamask-extension/commit/f5631fba1e83899f3e8d8a1bf2b1fc0de9fb87c6]

When a user has a transaction (STX) in a Pending state and initiates a transaction through a dApp on a different network, the dApp transaction is generated with an empty nonce. Pressing the 'Confirm' button then leads to an endless loading spinner. After reloading the extension, the second transaction's popup becomes accessible, and the transaction is then displayed with the correct nonce.

### Expected behavior

-

### Screenshots/Recordings

https://github.com/MetaMask/metamask-extension/assets/104780023/a24fda33-25a3-4ad0-b1c0-1dc422065279

### Steps to reproduce

1. Turn on nonce customization in the SEttings->Advanced
2. Connect to Ethereum
3. Send STX (set low gas fee so the STX stays in 'Pending' state for awhile)
4. Connect to dapp (Pancake Swap, Uniswap) to BNB or other network 
5. Start a dapp transaction
6. Notice that nonce field is empty
7. Confirm the transaction, see the loading spinner

### Error messages or log output

```shell

```

### Version

v12.0.0

### Build type

None

### Browser

Chrome

### Operating system

MacOS

### Hardware wallet

_No response_

### Additional context

_No response_

### Severity

_No response_

### PR DESCRIPTION
**Pull request by vinistevam**
## **Description**

This PR aims to fix the issue when a user has a Smart transaction in a pending state and initiates a new transaction through a dApp on a different network, the dApp transaction is generated with an empty nonce.

The issue happens due to a limitation on nonce management in the Transaction Controller. The nonce locks for the first transaction (STX) and only allows the user to submit another transaction once the first one is confirmed. This is the current behaviour for STX with transactions in the same network. This update extends the validation across different networks. 

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/25629?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/25579

## **Manual testing steps**

1. Turn on nonce customization in the SEttings->Advanced
2. Connect to Ethereum
3. Send STX (set low gas fee so the STX stays in 'Pending' state for awhile)
4. Connect to dapp (Pancake Swap, Uniswap) to BNB or other network 
5. Start a dapp transaction
6. Notice that nonce field is empty
7. Confirm the transaction, see the loading spinner

## **Screenshots/Recordings**

[Screencast from 02-07-2024 14:05:18.webm](https://github.com/MetaMask/metamask-extension/assets/45455812/2bded87f-6c66-4f78-acaf-badf41ce4563)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by sleepytanya** [comment]
User cannot send another transaction while having STX in `Pending` state. `Previous transaction is still being signed or submitted `message is shown and `Confirm` button is disabled.



https://github.com/MetaMask/metamask-extension/assets/104780023/e063dc24-b09f-4b03-b656-a34ad1ed85b4

**Comment by sleepytanya** [comment]
@vinistevam @bschorchit 

Should all types of txs be disabled if user has Pending STX?

Confusing UX when the first STX is of Approval type.
1. Send Approval STX on mainnet
2. Start transaction on another network
3. MM popup shows Submitting your transaction STX modal (user can't see the second transaction - which is expected)
4. Close the popup
5. When Approval STX is confirmed the popup is not visible (stays behind the browser window - this is another bug), user have to look for the popup / click on the blue badge again / reject, confirm from the Activity tab.


https://github.com/MetaMask/metamask-extension/assets/104780023/5a4551f3-7118-4932-97b1-e5e8a8936c2c



The error in the console when the STX gets confirmed:

Screenshot 2024-07-02 at 15 24 04

**Review by pedronfigueiredo — DISMISSED** [review]

File: ui/selectors/transactions.js [hunk 1]
```diff
@@ -36,6 +36,19 @@ const allowedSwapsSmartTransactionStatusesForActivityList = [
...
+    if (!transactions?.length) {
+      return [];
+    }
+
+    return transactions.sort((a, b) => a.time - b.time); // Ascending
+  },
+  (transactions) => transactions,
+);
+
 export const getCurrentNetworkTransactions = createDeepEqualSelector(
```
Comment (matthewwalsh0) [inline ui/selectors/transactions.js:52]: Should we change this to use `getTransactions` as a dependency to avoid the duplicated sorting?

**Review by matthewwalsh0 — DISMISSED** [review]

File: ui/selectors/transactions.js (hunk 1 above)
Comment (vinistevam) [reply ui/selectors/transactions.js:52]: I tried that yesterday but some of the unit tests started to fail. After your comment, I tried again, found the issue in the unit tests, and pushed this change to use `getTransactions`.

**Review by pedronfigueiredo — APPROVED** [review]

**Review by matthewwalsh0 — APPROVED** [review]

**Pull request merged**